spring-cloud-alibaba-starters/spring-cloud-alibaba-governance/governance-core/spring-cloud-starter-alibaba-governance-labelrouting/label-routing-data/src/main/java/com/alibaba/cloud/router/data/repository/RouteDataRepository.java

```diff
@@ -16,113 +16,247 @@
+	public static final int SUM_WEIGHT = 100;
+
+	/**
+	 * Weight value can't less than it.
+	 */
+	public static final int MIN_WEIGHT = 0;
+
+	public void init(final List<UntiedRouteDataStructure> routerDataList) {
+		int initCacheSize = routerDataList.size();
+		routeCache = new ConcurrentHashMap<>(initCacheSize);
+		originalRouteData = new ConcurrentHashMap<>(initCacheSize);
+		pathRuleMap = new ConcurrentHashMap<>(initCacheSize);
 		putRouteData(routerDataList);
 	}
 
-	public void updateRouteData(List<UntiedRouteDataStructure> routerDataList) {
+	public void updateRouteData(final List<UntiedRouteDataStructure> routerDataList) {
 		this.routeDataList = routerDataList;
 		routeDataChanged = true;
-
 		updateRouteData();
 	}
 
-	private void updateRouteData() {
+	/**
+	 * @return Finished updating data index.
+	 */
+	private int updateRouteData() {
 		while (routeDataChanged) {
 			int routeDataListSize = routeDataList.size();
-
+			// If all tasks had been distributed
 			if (waitUpdateIndex.get() == routeDataListSize) {
-				return;
+				return updateIndex.get();
 			}
-
+			// If not,get a task.
 			int i = waitUpdateIndex.incrementAndGet();
-
-			// avoid generate critical condition.
+			// May be multi-thread competition,avoid critical condition.
 			if (i < routeDataListSize) {
 				UntiedRouteDataStructure routerData = routeDataList.get(i);
-				LabelRouteData labelRouteData = routeCache
+				nonNullCheck(routerData);
+				LabelRouteData labelRouteData = originalRouteData
 						.get(routerData.getTargetService());
 
 				if (!routerData.getLabelRouteData().equals(labelRouteData)) {
-					putRouteData(routerData);
+					buildHashIndex(routerData);
+					originalRouteData.put(routerData.getTargetService(),
+							routerData.getLabelRouteData());
 				}
 				int updateNumber = updateIndex.incrementAndGet();
 
-				if (updateNumber >= routeDataListSize) {
+				if (updateNumber >= routeDataListSize - 1) {
 					routeDataChanged = false;
+					waitUpdateIndex.set(-1);
+					updateIndex.set(-1);
 				}
 			}
 		}
-	}
 
-	private void putRouteData(UntiedRouteDataStructure routerData) {
-		LabelRouteData putLabelRouteData = routeCache.put(routerData.getTargetService(),
-				routerData.getLabelRouteData());
+		return updateIndex.get();
 	}
 
-	private void putRouteData(List<UntiedRouteDataStructure> routerDataList) {
-		LabelRouteData putLabelRouteData = null;
-
-		for (UntiedRouteDataStructure routerData : routerDataList) {
-			putLabelRouteData = routeCache.put(routerData.getTargetService(),
-					routerData.getLabelRouteData());
-			if (putLabelRouteData == null) {
-				log.info("Label route rule:" + routerData
-						+ "had been add to router cache");
-			}
-		}
-	}
-
-	public LabelRouteData getRouteData(String targetService) {
-		// double check.
+	private void updateData(String targetService) {
 		while (routeDataChanged) {
-			updateRouteData();
-
+			// Update label rule data.
+			int updateIndex = updateRouteData();
+			// Double check.
```

Comment:
I don't understand here. If you want to solve thread safe problem, the method inner variable `routeDataChanged` can't seen in multh-thread and it can't achieve double check in different thread. It must be a Class variable by using keyword of `static`.

Reply:
logic has optimize